# Data Cleaning

In [18]:
import pandas as pd
import numpy as np

In [19]:
zomato=pd.read_csv("india1.csv")

In [20]:
zomato.head()

res_id                            name      establishment  \
0  3400299                     Bikanervala    ['Quick Bites']   
1  3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2  3401013                   Bhagat Halwai    ['Quick Bites']   
3  3400290                   Bhagat Halwai    ['Quick Bites']   
4  3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   

                                                 url  \
0  https://www.zomato.com/agra/bikanervala-khanda...   
1  https://www.zomato.com/agra/mama-chicken-mama-...   
2  https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3  https://www.zomato.com/agra/bhagat-halwai-civi...   
4  https://www.zomato.com/agra/the-salt-cafe-kitc...   

                                             address  city  city_id  \
0  Kalyani Point, Near Tulsi Cinema, Bypass Road,...  Agra       34   
1        Main Market, Sadar Bazaar, Agra Cantt, Agra  Agra       34   
2  62/1, Near Easy Day, West Shivaji Nagar, Goalp...  Agra       34   
3  Near Anjana Cinema, Nehru Nagar, Civil Lines, ...  Agra       34   
4        1C,3rd Floor, Fatehabad Road, Tajganj, Agra  Agra       34   

      locality   latitude  longitude  ... price_range  currency  \
0     Khandari  27.211450  78.002381  ...           2       Rs.   
1   Agra Cantt  27.160569  78.011583  ...           2       Rs.   
2     Shahganj  27.182938  77.979684  ...           1       Rs.   
3  Civil Lines  27.205668  78.004799  ...           1       Rs.   
4      Tajganj  27.157709  78.052421  ...           3       Rs.   

                                          highlights aggregate_rating  \
0  ['Lunch', 'Takeaway Available', 'Credit Card',...              4.4   
1  ['Delivery', 'No Alcohol Available', 'Dinner',...              4.4   
2  ['No Alcohol Available', 'Dinner', 'Takeaway A...              4.2   
3  ['Takeaway Available', 'Credit Card', 'Lunch',...              4.3   
4  ['Lunch', 'Serves Alcohol', 'Cash', 'Credit Ca...              4.9   

  rating_text  votes  photo_count opentable_support delivery  takeaway  
0   Very Good    814          154               0.0       -1        -1  
1   Very Good   1203          161               0.0       -1        -1  
2   Very Good    801          107               0.0        1        -1  
3   Very Good    693          157               0.0        1        -1  
4   Excellent    470          291               0.0        1        -1  

[5 rows x 26 columns]

In [21]:
zomato.shape

(211944, 26)

In [22]:
zomato.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211944 entries, 0 to 211943
Data columns (total 26 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   res_id                211944 non-null  int64  
 1   name                  211944 non-null  object 
 2   establishment         211944 non-null  object 
 3   url                   211944 non-null  object 
 4   address               211810 non-null  object 
 5   city                  211944 non-null  object 
 6   city_id               211944 non-null  int64  
 7   locality              211944 non-null  object 
 8   latitude              211944 non-null  float64
 9   longitude             211944 non-null  float64
 10  zipcode               48757 non-null   object 
 11  country_id            211944 non-null  int64  
 12  locality_verbose      211944 non-null  object 
 13  cuisines              210553 non-null  object 
 14  timings               208070 non-null  object 
 15  

In [23]:
zomato.describe()

res_id        city_id       latitude      longitude  country_id  \
count  2.119440e+05  211944.000000  211944.000000  211944.000000    211944.0   
mean   1.349411e+07    4746.785434      21.499758      77.615276         1.0   
std    7.883722e+06    5568.766386      22.781331       7.500104         0.0   
min    5.000000e+01       1.000000       0.000000       0.000000         1.0   
25%    3.301027e+06      11.000000      15.496071      74.877961         1.0   
50%    1.869573e+07      34.000000      22.514494      77.425971         1.0   
75%    1.881297e+07   11306.000000      26.841667      80.219323         1.0   
max    1.915979e+07   11354.000000   10000.000000      91.832769         1.0   

       average_cost_for_two    price_range  aggregate_rating          votes  \
count         211944.000000  211944.000000     211944.000000  211944.000000   
mean             595.812229       1.882535          3.395937     378.001864   
std              606.239363       0.892989          1.283642     925.333370   
min                0.000000       1.000000          0.000000     -18.000000   
25%              250.000000       1.000000          3.300000      16.000000   
50%              400.000000       2.000000          3.800000     100.000000   
75%              700.000000       2.000000          4.100000     362.000000   
max            30000.000000       4.000000          4.900000   42539.000000   

         photo_count  opentable_support       delivery  takeaway  
count  211944.000000           211896.0  211944.000000  211944.0  
mean      256.971224                0.0      -0.255907      -1.0  
std       867.668940                0.0       0.964172       0.0  
min         0.000000                0.0      -1.000000      -1.0  
25%         3.000000                0.0      -1.000000      -1.0  
50%        18.000000                0.0      -1.000000      -1.0  
75%       128.000000                0.0       1.000000      -1.0  
max     17702.000000                0.0       1.000000      -1.0

Removing duplicates
Its important to remove duplicate rows to avoid biasness in our analysis. Since res_id is unique identifier of our restaurants, we can use it to remove duplicates.

In [24]:
zomato.drop_duplicates(["res_id"],keep='first',inplace=True)
zomato.shape

(55568, 26)

missig value

In [25]:
zomato.isnull().sum()

res_id                      0
name                        0
establishment               0
url                         0
address                    18
city                        0
city_id                     0
locality                    0
latitude                    0
longitude                   0
zipcode                 44623
country_id                  0
locality_verbose            0
cuisines                  470
timings                  1003
average_cost_for_two        0
price_range                 0
currency                    0
highlights                  0
aggregate_rating            0
rating_text                 0
votes                       0
photo_count                 0
opentable_support          12
delivery                    0
takeaway                    0
dtype: int64

We have 5 variables with some kind of missing values. Since zipcode has ~80% missing data, its better to not consider it at all. The other 4 features can be delt with some kind of imputation, but before going through the trouble, its better to look and decide whether they would be beneficial for our analysis or we can simply omit them

Omitting not useful features

Here we will look at each feature and decide to consider them for our analysis or not:-

res_id - Unique ID for each restaurant

name - Name is useful since we will use it to find top restaurants

establishment - Let's see what type of values we have in establishment

In [26]:
zomato["establishment"].unique()

array(["['Quick Bites']", "['Casual Dining']", "['Bakery']", "['Café']",
       "['Dhaba']", "['Bhojanalya']", "['Bar']", "['Sweet Shop']",
       "['Fine Dining']", "['Food Truck']", "['Dessert Parlour']",
       "['Lounge']", "['Pub']", "['Beverage Shop']", "['Kiosk']",
       "['Paan Shop']", "['Confectionery']", '[]', "['Shack']",
       "['Club']", "['Food Court']", "['Mess']", "['Butcher Shop']",
       "['Microbrewery']", "['Cocktail Bar']", "['Pop up']",
       "['Irani Cafe']"], dtype=object)

In [27]:
print(zomato["establishment"].unique()[0])
print(type(zomato["establishment"].unique()[0]))

['Quick Bites']
<class 'str'>


each value has an unwanted square brackets and quotes which seems noisy. Let's remove them with apply( ) function. Also, we have one value which is an empty string, let's rename it to "NA" to avoid confusion.

In [30]:
# Removing [' '] from each value
print(zomato["establishment"].unique()[0])
zomato["establishment"] = zomato["establishment"].apply(lambda x:x[2:-2])
print(zomato["establishment"].unique()[0])

# Changing ''  to 'NA'
print(zomato["establishment"].unique())
zomato["establishment"] = zomato["establishment"].apply(lambda x : np.where(x=="", "NA", x))
print(zomato["establishment"].unique())

['Quick Bites']
Quick Bites
['Quick Bites' 'Casual Dining' 'Bakery' 'Café' 'Dhaba' 'Bhojanalya' 'Bar'
 'Sweet Shop' 'Fine Dining' 'Food Truck' 'Dessert Parlour' 'Lounge' 'Pub'
 'Beverage Shop' 'Kiosk' 'Paan Shop' 'Confectionery' '' 'Shack' 'Club'
 'Food Court' 'Mess' 'Butcher Shop' 'Microbrewery' 'Cocktail Bar' 'Pop up'
 'Irani Cafe']
['Quick Bites' 'Casual Dining' 'Bakery' 'Café' 'Dhaba' 'Bhojanalya' 'Bar'
 'Sweet Shop' 'Fine Dining' 'Food Truck' 'Dessert Parlour' 'Lounge' 'Pub'
 'Beverage Shop' 'Kiosk' 'Paan Shop' 'Confectionery' 'NA' 'Shack' 'Club'
 'Food Court' 'Mess' 'Butcher Shop' 'Microbrewery' 'Cocktail Bar' 'Pop up'
 'Irani Cafe']


1. url - URL is the link to restaurant's page which is not useful for us
2. address - Not useful since it has long strings and its difficult to classify
3. city - Let's check unique cities

In [31]:
zomato["city"].unique()

array(['Agra', 'Ahmedabad', 'Gandhinagar', 'Ajmer', 'Alappuzha',
       'Allahabad', 'Amravati', 'Amritsar', 'Aurangabad', 'Bangalore',
       'Bhopal', 'Bhubaneshwar', 'Chandigarh', 'Mohali', 'Panchkula',
       'Zirakpur', 'Nayagaon', 'Chennai', 'Coimbatore', 'Cuttack',
       'Darjeeling', 'Dehradun', 'New Delhi', 'Gurgaon', 'Noida',
       'Faridabad', 'Ghaziabad', 'Greater Noida', 'Dharamshala',
       'Gangtok', 'Goa', 'Gorakhpur', 'Guntur', 'Guwahati', 'Gwalior',
       'Haridwar', 'Hyderabad', 'Secunderabad', 'Indore', 'Jabalpur',
       'Jaipur', 'Jalandhar', 'Jammu', 'Jamnagar', 'Jamshedpur', 'Jhansi',
       'Jodhpur', 'Junagadh', 'Kanpur', 'Kharagpur', 'Kochi', 'Kolhapur',
       'Kolkata', 'Howrah', 'Kota', 'Lucknow', 'Ludhiana', 'Madurai',
       'Manali', 'Mangalore', 'Manipal', 'Udupi', 'Meerut', 'Mumbai',
       'Thane', 'Navi Mumbai', 'Mussoorie', 'Mysore', 'Nagpur',
       'Nainital', 'Nasik', 'Nashik', 'Neemrana', 'Ooty', 'Palakkad',
       'Patiala', 'Patna', 'Pudu

Look's good.

1. city_id - We can uniquely use city name or id. So one feature is enough
2. locality - Let's see number of unique values

In [32]:
zomato["locality"].nunique()

3731

Although it can be an interesting feature, but since this feature has so many unique classes, we will avoid it.

1. latitude - Can be helpful while using geographic maps
2. longitude - Can be helpful while using geographic maps
3. zipcode - Approx 80% missing values
4. country_id - Since this dataset is for Indian restaurants, there should be just one unique id here. Let's check.

In [34]:
zomato["country_id"].unique()

array([1], dtype=int64)

### locality_verbose - Same as locality

In [36]:
zomato["locality_verbose"].nunique()

3910

1. cuisines - This feature has some missing values. Even though this has 9382 unique classes, we can see that each restaurant has a list of cusinies and the composition of the list is the reason why we have so many different cuisine classes. Let's check actual number of unique cuisine classes. But first we need to replace null values with a label.

In [37]:
print(zomato["cuisines"].nunique())
print(zomato["cuisines"].unique())

9382
['North Indian, South Indian, Mithai, Street Food, Desserts'
 'North Indian, Mughlai, Rolls, Chinese, Fast Food, Street Food'
 'Fast Food, Mithai' ...
 'Street Food, Biryani, Chinese, Fast Food, North Indian, Mughlai'
 'North Indian, Chinese, Mexican, Italian, Thai, Continental'
 'North Indian, Lucknowi, Chinese']


In [38]:
zomato["cuisines"] = zomato["cuisines"].fillna("No cuisine")

In [39]:
cuisines = []
zomato["cuisines"].apply(lambda x : cuisines.extend(x.split(", ")))
cuisines = pd.Series(cuisines)
print("Total number of unique cuisines = ", cuisines.nunique())

Total number of unique cuisines =  134


1. timings - This also has missing data, however it has 7740 unique classes. Also, it is not structured even if we try to reduce the number classes like we did in cuisines. Its better to omit it altogether.

In [40]:
print(zomato["timings"].nunique())
print(zomato["timings"].unique())

7740
['8:30am – 10:30pm (Mon-Sun)'
 '12:30PM to 12Midnight (Mon, Wed, Thu, Fri, Sat, Sun), 1PM to 12Midnight (Tue)'
 '9:30 AM to 11 PM' ... '8am – 2:30pm, 5pm – 10pm (Mon-Sun)'
 '8am – 3pm, 6:30pm – 11pm (Mon-Sun)'
 '4pm – 11pm (Mon, Tue, Wed, Fri, Sat, Sun), 4pm – 11:30pm (Thu)']


1. average_cost_for_two - This is an interesting feature for our analysis, although the value "0" is strange and should be an outlier

In [41]:
zomato["average_cost_for_two"].unique()

array([  700,   600,   300,  1000,   400,   500,  1200,  1600,   250,
         200,   150,  1500,   800,   450,   850,   350,   900,  6000,
         650,   550,  2500,  6500,  5500,  4300,  3600,  3500,  3200,
        3000,  2800,  2000,  1800,  1400,  1300,  1250,  1150,  1100,
        1050,    50,    60,   100,   750,     0,   220,   260,   950,
        1900,  2400,  2300,  2200,  1700,  1230,    70,   140,   120,
         470,    20,    40,   110,    80,    90,   440,  2100, 14000,
        5000,  4500,  4100,  4000,  3700,  3400,  2700,  2600,  1650,
        1350,   180,  1260,     2,   197,   312,   170,   240,   330,
        3100,  4200,  1850,  3300,  4800,  3800,  3250,  2900,  1750,
         282,    35,   149,  1550,  5100,  6600,  1450,   160,  2250,
        8000,  7000,  3900,  3650,  1020,   320,   360,   280,   130,
       15000,  2501,    30,  1950,  1875,  1190,   309, 12000,   380,
         299,    45,   190,  4600,   880,    75,   620,   999, 10000,
         505, 30000,

1. price_range - Average prices automatically characterized into bins

In [42]:
zomato["price_range"].unique()

array([2, 1, 3, 4], dtype=int64)

1. currency - Only one class. Not useful

In [43]:
zomato["currency"].unique()

array(['Rs.'], dtype=object)

1. highlights - They represent certain features that the restaurant specializes in and wants to highlight to their customers. Each restaurant has a list of highlights which makes the composition different for each one. We can, filter this and find total unique highlights from all restaurants.

In [44]:
print(zomato["highlights"].nunique())
print(zomato["highlights"].unique())

31228
["['Lunch', 'Takeaway Available', 'Credit Card', 'Dinner', 'Cash', 'Air Conditioned', 'Indoor Seating', 'Pure Veg']"
 "['Delivery', 'No Alcohol Available', 'Dinner', 'Takeaway Available', 'Lunch', 'Cash', 'Indoor Seating']"
 "['No Alcohol Available', 'Dinner', 'Takeaway Available', 'Breakfast', 'Lunch', 'Cash', 'Delivery', 'Outdoor Seating', 'Air Conditioned', 'Self Service', 'Indoor Seating', 'Digital Payments Accepted', 'Pure Veg', 'Desserts and Bakes']"
 ...
 "['Dinner', 'Delivery', 'Cash', 'Takeaway Available', 'Free Parking', 'Digital Payments Accepted', 'Pure Veg', 'Indoor Seating']"
 "['Dinner', 'Cash', 'Takeaway Available', 'Lunch', 'Delivery', 'Free Parking', 'Indoor Seating', 'Air Conditioned', 'Outdoor Seating', 'Digital Payments Accepted', 'Catering Available', 'Pure Veg']"
 "['Dinner', 'Cash', 'Takeaway Available', 'Debit Card', 'Delivery', 'Credit Card', 'Free Parking', 'Outdoor Seating']"]


In [45]:
hl = []
zomato["highlights"].apply(lambda x : hl.extend(x[2:-2].split("', '")))
hl = pd.Series(hl)
print("Total number of unique highlights = ", hl.nunique())

Total number of unique highlights =  104


1. aggregate_rating - Rating given to the restaurant
2. rating_text - Characterisation of numeric rating into bins by using labels. We will be using direct ratings in our analysis, so we can ignore this.
3. votes - Number of votes contributing to the rating
4. photo_count - Photo uploads in reviews
Let's check the mean and range of above features

In [46]:
zomato[["aggregate_rating","votes","photo_count"]].describe().loc[["mean","min","max"]]

aggregate_rating         votes  photo_count
mean          2.958593    223.330352    160.97477
min           0.000000    -18.000000      0.00000
max           4.900000  42539.000000  17702.00000

Rating ranges between 0 and 5 while 42539 are the maximum votes given to a restaurant. The negative value in votes might be an outlier.

1. opentable_support - Not useful since no restaurant has True value for this
2. delivery - This feature has 3 classes but there is no explanation for those classes. We can consider -1 and 0 to be one class or ignore this feature for now
3. takeaway - Again not useful since it only has one class

In [47]:
zomato["opentable_support"].unique()

array([ 0., nan])

In [48]:
zomato["delivery"].unique()

array([-1,  1,  0], dtype=int64)

In [50]:
zomato["takeaway"].unique()

array([-1], dtype=int64)

In [55]:
zomato.tail()

res_id                   name    establishment  \
211882  19142822  Shree Janta Ice Cream  Dessert Parlour   
211925  18984164       The Grand Thakar    Casual Dining   
211926  18019952            Geeta lodge    Casual Dining   
211940   3200996             Raju Omlet      Quick Bites   
211942   3201138                 Subway      Quick Bites   

                                                      url  \
211882  https://www.zomato.com/vadodara/shree-janta-ic...   
211925  https://www.zomato.com/vadodara/the-grand-thak...   
211926  https://www.zomato.com/vadodara/geeta-lodge-al...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   

                                                  address      city  city_id  \
211882  Ground Floor, 5 Ronak Plaza, Tulsidham Char Ra...  Vadodara       32   
211925  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211926  Shop 11, Ground Floor, Atlantis K-10, Tower A,...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   

          locality   latitude  longitude  ... price_range  currency  \
211882   Manjalpur  22.270516  73.196408  ...           1       Rs.   
211925    Alkapuri  22.310563  73.171163  ...           2       Rs.   
211926    Alkapuri  22.317731  73.168107  ...           1       Rs.   
211940  Karelibaug  22.322455  73.197203  ...           1       Rs.   
211942       Akota  22.270027  73.143068  ...           2       Rs.   

                                               highlights aggregate_rating  \
211882  ['Cash', 'Takeaway Available', 'Delivery', 'In...              2.9   
211925  ['Dinner', 'Cash', 'Debit Card', 'Lunch', 'Tak...              4.0   
211926  ['Dinner', 'Cash', 'Credit Card', 'Lunch', 'Ta...              3.9   
211940  ['Dinner', 'Cash', 'Takeaway Available', 'Debi...              4.1   
211942  ['Dinner', 'Delivery', 'Credit Card', 'Lunch',...              3.7   

       rating_text  votes  photo_count opentable_support delivery  takeaway  
211882     Average      4            1               0.0        1        -1  
211925   Very Good    111           38               0.0       -1        -1  
211926        Good    207           14               0.0       -1        -1  
211940   Very Good    187           40               0.0        1        -1  
211942        Good    128           34               0.0        1        -1  

[5 rows x 26 columns]

In [57]:
zomato.drop(columns=["url","address","opentable_support","currency","timings","zipcode","country_id"],axis=1,inplace=True)

In [58]:
zomato.head()

res_id                            name  establishment  city  city_id  \
0  3400299                     Bikanervala    Quick Bites  Agra       34   
1  3400005  Mama Chicken Mama Franky House    Quick Bites  Agra       34   
2  3401013                   Bhagat Halwai    Quick Bites  Agra       34   
3  3400290                   Bhagat Halwai    Quick Bites  Agra       34   
4  3401744     The Salt Cafe Kitchen & Bar  Casual Dining  Agra       34   

      locality   latitude  longitude   locality_verbose  \
0     Khandari  27.211450  78.002381     Khandari, Agra   
1   Agra Cantt  27.160569  78.011583   Agra Cantt, Agra   
2     Shahganj  27.182938  77.979684     Shahganj, Agra   
3  Civil Lines  27.205668  78.004799  Civil Lines, Agra   
4      Tajganj  27.157709  78.052421      Tajganj, Agra   

                                            cuisines  average_cost_for_two  \
0  North Indian, South Indian, Mithai, Street Foo...                   700   
1  North Indian, Mughlai, Rolls, Chinese, Fast Fo...                   600   
2                                  Fast Food, Mithai                   300   
3          Desserts, Bakery, Fast Food, South Indian                   300   
4                 North Indian, Continental, Italian                  1000   

   price_range                                         highlights  \
0            2  ['Lunch', 'Takeaway Available', 'Credit Card',...   
1            2  ['Delivery', 'No Alcohol Available', 'Dinner',...   
2            1  ['No Alcohol Available', 'Dinner', 'Takeaway A...   
3            1  ['Takeaway Available', 'Credit Card', 'Lunch',...   
4            3  ['Lunch', 'Serves Alcohol', 'Cash', 'Credit Ca...   

   aggregate_rating rating_text  votes  photo_count  delivery  takeaway  
0               4.4   Very Good    814          154        -1        -1  
1               4.4   Very Good   1203          161        -1        -1  
2               4.2   Very Good    801          107         1        -1  
3               4.3   Very Good    693          157         1        -1  
4               4.9   Excellent    470          291         1        -1

In [60]:
zomato.to_csv('./zomato.csv')